geo: order geospatial objects by Hilbert Curve index

otan opened the pull request:
This follows the [PostGIS way](https://info.crunchydata.com/blog/waiting-for-postgis-3-hilbert-geometry-sorting),
but does not follow the same encoding.

Visualisation for Geometry for 10000 random points: 
![image](https://user-images.githubusercontent.com/3646147/88688829-9a4f5a00-d0ae-11ea-9618-0179fb7ac327.png)

Visualisation for Geography  10000 random points:
![image](https://user-images.githubusercontent.com/3646147/88449411-93e58780-cdfb-11ea-9e85-eff8c5a94787.png)

Release note (sql change): When ordering by geospatial columns, it will
now order by the Hilbert Space Curve index so that points which are
geographically similar are clustered together.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51898)
<!-- Reviewable:end -->

mjibson commented on pkg/geo/geo.go:
```diff
@@ -276,6 +278,52 @@ func (g *Geometry) CartesianBoundingBox() *CartesianBoundingBox {
 	return &CartesianBoundingBox{BoundingBox: *g.spatialObject.BoundingBox}
 }
 
+// SpaceCurveIndex returns an uint64 index to use representing an index into a Space Curve.
```
Add some comments here about what happens in error or edge cases (like return 0 or math.MaxUint64). Also document a bit of what a Space Curve means, like, can you compare them? Is there a collision risk such that some objects will return the same value even though they are not actually equal? How does this function relate to CompareSpatialObject?

otan replied:
Done

mjibson commented on pkg/geo/geo.go:
```diff
@@ -489,6 +537,15 @@ func (g *Geography) BoundingCap() s2.Cap {
 	return g.BoundingRect().CapBound()
 }
 
+// SpaceCurveIndex returns an uint64 index to use representing an index into a Space Curve.
```
More comments, as above.

madelynnblue reviewed (commented):
There's now CompareSpatialObject and SpaceCurveIndex. It is then up to callers in other packages that they need to call both in a specific order. Should we instead make a single function in this package that does the correct thing? If not, we should document somewhere the relationship between those two functions and how they should be used when comparing objects.

mjibson commented on pkg/util/encoding/encoding.go:
```diff
@@ -978,39 +978,55 @@ func decodeTime(b []byte) (r []byte, sec int64, nsec int64, err error) {
 	var ret geopb.SpatialObject
-	var err error
-	b, pbBytes, err = decodeBytesInternal(b, pbBytes, ascendingGeoEscapes, true)
+	b, pbBytes, err = decodeBytesInternal(b, pbBytes, ascendingGeoEscapes, false)
```
add a comment here for what the boolean argument is

mjibson commented on pkg/util/encoding/encoding.go:
```diff
@@ -1021,10 +1037,19 @@ func DecodeGeoAscending(b []byte) ([]byte, geopb.SpatialObject, error) {
 	var ret geopb.SpatialObject
-	var err error
-	b, pbBytes, err = decodeBytesInternal(b, pbBytes, descendingGeoEscapes, true)
+	b, pbBytes, err = decodeBytesInternal(b, pbBytes, descendingGeoEscapes, false)
```
add a comment here for what the boolean argument is

madelynnblue reviewed (commented):
Should be able to remove the geo exceptions in sqlbase/encoded_datum_test.go so that TestEncDatumCompare tests these now.

mjibson commented on pkg/util/encoding/encoding.go:
```diff
@@ -1527,13 +1552,21 @@ func PeekLength(b []byte) (int, error) {
 		return getBytesLength(b, ascendingBytesEscapes)
 	case geoMarker:
-		return getBytesLength(b, ascendingGeoEscapes)
+		ret, err := getBytesLength(b[8:], ascendingGeoEscapes)
```
You have not verified that `len(b) >= 8` here, so this could panic.

mjibson commented:
Also someone else should review the implementation of the new geo stuff here. I'm not qualified.

sumeerbhola reviewed (changes requested):
> Reviewed 3 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51898)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson, @otan, and @sumeerbhola)

___
*[pkg/geo/geo.go, line 281 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDGl9Uq42DZwTPawCkE:-MDGl9Uq42DZwTPawCkF:b5zku7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/6252e1f77b51924a31876f3514753cf323bf1cdb/pkg/geo/geo.go#L281)):*
> ```Go
> }
> 
> // SpaceCurveIndex returns an uint64 index to use representing an index into a Space Curve.
> ```

nit: ... an index into a Space-filling Curve.
___
*[pkg/geo/geo.go, line 291 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDGm19x5N0mcjNsAifH:-MDGm19x5N0mcjNsAifI:b-4i1068) ([raw file](https://github.com/cockroachdb/cockroach/blob/6252e1f77b51924a31876f3514753cf323bf1cdb/pkg/geo/geo.go#L291)):*
> ```Go
> 	centerX := (bbox.BoundingBox.LoX + bbox.BoundingBox.HiX) / 2
> 	centerY := (bbox.BoundingBox.LoY + bbox.BoundingBox.HiY) / 2
> 	// By default, bound by MaxFloat32.
> ```

The comment says `MaxFloat32` but the code is doing `MaxInt32`.
___
*[pkg/geo/geo.go, line 307 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDGng2W8j-0Gr1em95l:-MDGng2W8j-0Gr1em95m:bjx9jd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6252e1f77b51924a31876f3514753cf323bf1cdb/pkg/geo/geo.go#L307)):*
> ```Go
> 
> 	// Find the longest dimension, add one to be inclusive.
> 	// TODO(otan): investigate storing this calculation on the projection itself for perf.
> ```

How about a comment like:
// We need two compute N such that N is a power of 2 and the x, y coordinates are in [0, N-1].
// The hilbert curve value will be in the interval [0, N^2-1]. The x, y coordinates will be first normalized
// below to the interval [0, 1].
(the rest of the comment is a TODO, depending on the questions below)
___
*[pkg/geo/geo.go, line 309 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDGmv8j5cuiTQNagLmZ:-MDGmv8j5cuiTQNagLm_:bmxh1fs) ([raw file](https://github.com/cockroachdb/cockroach/blob/6252e1f77b51924a31876f3514753cf323bf1cdb/pkg/geo/geo.go#L309)):*
> ```Go
> 	// TODO(otan): investigate storing this calculation on the projection itself for perf.
> 	biggestBoxLength := math.Max(bounds.MaxX-bounds.MinX, bounds.MaxY-bounds.MinY) + 1
> 	boxSize := 1 << 40
> ```

what is the reason for this value `1 << 40` versus some other power of 2? Won't this overflow when doing `(2^40)^2-1`?

It is not clear why we are trying to compute a variable box size since we normalize to [0, 1] below before multiplying by the box size. Given the hilbert library is using signed `int`, it seems simpler to use a constant box size of `2^15`, so that we get a maximum return value of `2^30-1` (I didn't look at the implementation of the library to see what it does wrt overflow) which gives us the most ability to distinguish between different centers. 


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geo.go, line 309 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDGmv8j5cuiTQNagLmZ:-MDH0MfR8CByMulVUbQR:bzgjq1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/6252e1f77b51924a31876f3514753cf323bf1cdb/pkg/geo/geo.go#L309)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

what is the reason for this value `1 << 40` versus some other power of 2? Won't this overflow when doing `(2^40)^2-1`?

It is not clear why we are trying to compute a variable box size since we normalize to [0, 1] below before multiplying by the box size. Given the hilbert library is using signed `int`, it seems simpler to use a constant box size of `2^15`, so that we get a maximum return value of `2^30-1` (I didn't look at the implementation of the library to see what it does wrt overflow) which gives us the most ability to distinguish between different centers. 

</blockquote></details>

A signed int in Golang is int64 on 64-bit systems.
Using a constant size of 15 is not great -- this is the bounding.

![image.png](https://files.reviewableusercontent.io/images/reviewable/3646147/CkXcoKBljqyr_CYSQRE7xNuk/image.png)


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51898)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson and @sumeerbhola)

___
*[pkg/geo/geo.go, line 540 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MD3PA2hTdmaw0AyMpJ6-r1-540:-MDH2XWZD-4t-At2EGRt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c751ffaac4ebd9996edb7a4de7bcce67b8881401/pkg/geo/geo.go#L540)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

More comments, as above.
</blockquote></details>

Done.
___
*[pkg/geo/geo.go, line 291 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDGm19x5N0mcjNsAifH:-MDH02QtEDJ1eiSBjygY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6252e1f77b51924a31876f3514753cf323bf1cdb/pkg/geo/geo.go#L291)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The comment says `MaxFloat32` but the code is doing `MaxInt32`.
</blockquote></details>

Done.
___
*[pkg/geo/geo.go, line 307 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDGng2W8j-0Gr1em95l:-MDH2Vh991XtIR-NZNfy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6252e1f77b51924a31876f3514753cf323bf1cdb/pkg/geo/geo.go#L307)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

How about a comment like:
// We need two compute N such that N is a power of 2 and the x, y coordinates are in [0, N-1].
// The hilbert curve value will be in the interval [0, N^2-1]. The x, y coordinates will be first normalized
// below to the interval [0, 1].
(the rest of the comment is a TODO, depending on the questions below)
</blockquote></details>

Done.
___
*[pkg/util/encoding/encoding.go, line 1029 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MD3PA2hTdmaw0AyMpJA-r2-1029:-MDH2YH83iOHgh7IUmmR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41820b738a4d9d42679fe30d6d5c98ed83ca7f09/pkg/util/encoding/encoding.go#L1029)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

add a comment here for what the boolean argument is
</blockquote></details>

Done.
___
*[pkg/util/encoding/encoding.go, line 1052 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MD3PA2hTdmaw0AyMpJA-r2-1052:-MDH2Yjv0bwUP5-i_VsT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41820b738a4d9d42679fe30d6d5c98ed83ca7f09/pkg/util/encoding/encoding.go#L1052)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

add a comment here for what the boolean argument is
</blockquote></details>

Done.
___
*[pkg/util/encoding/encoding.go, line 1555 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MD3PA2hTdmaw0AyMpJA-r3-1555:-MDH2Y_762IS51ZH8PaR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f69c7019b0d4cd309b168ebf1fd7569e4dfa299/pkg/util/encoding/encoding.go#L1555)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

You have not verified that `len(b) >= 8` here, so this could panic.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geo.go, line 309 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDGmv8j5cuiTQNagLmZ:-MDH3t9S4NDa_cIK5cbb:batw2k8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6252e1f77b51924a31876f3514753cf323bf1cdb/pkg/geo/geo.go#L309)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

A signed int in Golang is int64 on 64-bit systems.
Using a constant size of 15 is not great -- this is the bounding.

![image.png](https://files.reviewableusercontent.io/images/reviewable/3646147/CkXcoKBljqyr_CYSQRE7xNuk/image.png)

</blockquote></details>

That said, this is for a 10000x10000 cell square for SRID 3857 (-20037508.342789244 to 20037508.342789244), so maybe this is acceptable?


<!-- Sent from Reviewable.io -->

mjibson commented on pkg/util/encoding/encoding_test.go:
```diff
@@ -1147,34 +1147,96 @@ func TestEncodeDecodeTimeTZ(t *testing.T) {
+	}
+	for _, dir := range []Direction{Ascending, Descending} {
+		t.Run(fmt.Sprintf("dir:%d", dir), func(t *testing.T) {
+			var lastEncoded []byte
```
This breaks subtests since you can no longer run just one. Is it annoying to use `testCases[i-1]` instead? I don't feel strongly, leave it as is if you want.

madelynnblue reviewed (approved):
Encoding stuff LGTM.

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 9 files at r1, 1 of 5 files at r2, 1 of 2 files at r3, 3 of 4 files at r5, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51898)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson, @otan, and @sumeerbhola)

___
*[pkg/geo/geo.go, line 309 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDGmv8j5cuiTQNagLmZ:-MDHqXng-QgaI_jOiTeU:bht6qye) ([raw file](https://github.com/cockroachdb/cockroach/blob/6252e1f77b51924a31876f3514753cf323bf1cdb/pkg/geo/geo.go#L309)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

That said, this is for a 10000x10000 cell square for SRID 3857 (-20037508.342789244 to 20037508.342789244), so maybe this is acceptable?
</blockquote></details>

let's write our own inverse function using `uint64` -- the algorithm is well documented https://en.wikipedia.org/wiki/Hilbert_curve#Applications_and_mapping_algorithms
___
*[pkg/util/encoding/encoding.go, line 1006 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDHnpgo5Ua0WaNkyqSd:-MDHnpgo5Ua0WaNkyqSe:b-8xm9hn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b56e3086639e253f46411bb1e19f8e212592e3dc/pkg/util/encoding/encoding.go#L1006)):*
> ```Go
> 	}
> 	n := len(b)
> 	b = encodeBytesAscendingWithTerminator(b, data, ascendingGeoEscapes.escapedTerm)
> ```

How about adding a comment
// These bytes don't provide a semantically meaningful order, unlike the curveIndex, so we encode these in ascending order even though this is the `EncodeGeoDescending` function. 

Or maybe this was meant to be descending since I see `DecodeGeoDescending()` doing `descendingGeoEscapes`. If this was a typo, can you add a test that would catch it.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/util/encoding/encoding.go, line 1006 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDHnpgo5Ua0WaNkyqSd:-MDHt9ENCMiA5fbkyDjc:b-pewvlm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b56e3086639e253f46411bb1e19f8e212592e3dc/pkg/util/encoding/encoding.go#L1006)):*
These is a line below:

> 	onesComplement(b[n:])

that inverses this. I can add a comment here, but tests do catch it.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geo.go, line 309 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDGmv8j5cuiTQNagLmZ:-MDHtGWRCkHcmjquhwvC:b-cm9kq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6252e1f77b51924a31876f3514753cf323bf1cdb/pkg/geo/geo.go#L309)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

let's write our own inverse function using `uint64` -- the algorithm is well documented https://en.wikipedia.org/wiki/Hilbert_curve#Applications_and_mapping_algorithms
</blockquote></details>

Something I'm not quite understanding -- is there a reason to? 
I think this one is fine; it's just with a small resolution (i.e. 15), it doesn't go deep enough
Also if we're going to write our own, we could use https://github.com/rawrunprotected/hilbert_curves/ which is advertised as log(n), but I'm not sure I understand the math.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51898)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson, @otan, and @sumeerbhola)

___
*[pkg/geo/geo.go, line 309 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDGmv8j5cuiTQNagLmZ:-MDK6sxsEzukSGQmQGfp:bw69l69) ([raw file](https://github.com/cockroachdb/cockroach/blob/6252e1f77b51924a31876f3514753cf323bf1cdb/pkg/geo/geo.go#L309)):*
> Something I'm not quite understanding -- is there a reason to?
> I think this one is fine; it's just with a small resolution (i.e. 15), it doesn't go deep enough

It is a reasonable question. My opinion is that writing those 15 lines of code is faster than trying to debate or think through all the cons of the tradeoff we are making by limiting ourselves to 2^15 resolution for each coordinate. And your visualization with the intersecting lines in the lower resolution setting look quite compellingly bad.

Regarding the actual algorithm, I'd rather go with the common one.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51898)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson and @sumeerbhola)

___
*[pkg/geo/geo.go, line 309 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDGmv8j5cuiTQNagLmZ:-MDM-lq6EHXQxjw_9ZRe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6252e1f77b51924a31876f3514753cf323bf1cdb/pkg/geo/geo.go#L309)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> Something I'm not quite understanding -- is there a reason to?
> I think this one is fine; it's just with a small resolution (i.e. 15), it doesn't go deep enough

It is a reasonable question. My opinion is that writing those 15 lines of code is faster than trying to debate or think through all the cons of the tradeoff we are making by limiting ourselves to 2^15 resolution for each coordinate. And your visualization with the intersecting lines in the lower resolution setting look quite compellingly bad.

Regarding the actual algorithm, I'd rather go with the common one.

</blockquote></details>

Done.
___
*[pkg/util/encoding/encoding.go, line 1006 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDHnpgo5Ua0WaNkyqSd:-MDM-jvk8nbLMhiqTP-i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b56e3086639e253f46411bb1e19f8e212592e3dc/pkg/util/encoding/encoding.go#L1006)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

These is a line below:

> 	onesComplement(b[n:])

that inverses this. I can add a comment here, but tests do catch it.
</blockquote></details>

Done.
___
*[pkg/util/encoding/encoding_test.go, line 1208 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MD3PA2ii9bb334zZ_i4-r5-1208:-MDM-kBs8AcmT2X6unQW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aef51d7fa4590f3d52f582d585b4c526491442a7/pkg/util/encoding/encoding_test.go#L1208)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

This breaks subtests since you can no longer run just one. Is it annoying to use `testCases[i-1]` instead? I don't feel strongly, leave it as is if you want.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 7 of 7 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51898)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson, @otan, and @sumeerbhola)

___
*[pkg/geo/geo.go, line 309 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDQXeCbCQk8OBolY-0C:-MDQXeCbCQk8OBolY-0D:b-a0eax8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d2dd6e52b6b67e75f12201d79f50d702d7a9a4c/pkg/geo/geo.go#L309)):*
> ```Go
> 	yBounds := (bounds.MaxY - bounds.MinY) + 1
> 
> 	// We need two compute N such that N is a power of 2 and the x, y coordinates are in [0, N-1].
> ```

s/two/to/

The logic in this function seems a bit confusing and I am not sure about its correctness:
- it isn't quite obvious why the `xBounds` and `yBounds` equation above ensures that the `xPos,yPos` computed later will be `< N`. If `xBounds` is a power of 2 and `((centerX - bounds.MinX) / (bounds.MaxX - bounds.MinX))` is 1, then `xPos=N` I think.

- If the bounds are coming from the SRID, I don't see why they can't exceed the size of `uint32` which is undesirable since the `biggestBoxLength` could exceed `uint32`.

I think something like the following would work since it uses the bounds only for normalizing to [0, 1). Once that normalization is done, we can use a constant sized square for the hilbert curve with a side of 2^32.
```
	const boxLength = 1 << 32
	// Add 1 to each bound so that we normalize the coordinates to [0, 1) before
	// multiplying by boxLength to give coordinates that are integers in the interval [0, boxLength-1].
	xBounds := (bounds.MaxX - bounds.MinX) + 1
	yBounds := (bounds.MaxY - bounds.MinY) + 1
	xPos := uint64(((centerX - bounds.MinX) / xBounds) * boxLength)
	yPos := uint64(((centerY - bounds.MinY) / yBounds) * boxLength)
        return hilbertInverse(boxLength, xPos, yPos)
```
___
*[pkg/geo/geo_test.go, line 565 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDQzwYm0So6gBr2qxdj:-MDQzwYm0So6gBr2qxdk:bb6v7sk) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d2dd6e52b6b67e75f12201d79f50d702d7a9a4c/pkg/geo/geo_test.go#L565)):*
> ```Go
> }
> 
> func TestGeographyHash(t *testing.T) {
> ```

why hash?
___
*[pkg/util/encoding/encoding.go, line 982 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDR-uhu19t-b-EYOlfc:-MDR-uhv6NJVOt72JGsu:b-vemdol) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d2dd6e52b6b67e75f12201d79f50d702d7a9a4c/pkg/util/encoding/encoding.go#L982)):*
> ```Go
> // returns the new buffer.
> // It is sorted by the given curve index, followed by the bytes of the spatial object.
> func EncodeGeoAscending(b []byte, curveIndex uint64, g *geopb.SpatialObject) ([]byte, error) {
> ```

is this function used for encoding a geometry column even when it is not part of a key (when the ordering does not matter)?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51898)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson and @sumeerbhola)

___
*[pkg/geo/geo.go, line 309 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDQXeCbCQk8OBolY-0C:-MDRDD7LE8HnXhFYc9g9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d2dd6e52b6b67e75f12201d79f50d702d7a9a4c/pkg/geo/geo.go#L309)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

s/two/to/

The logic in this function seems a bit confusing and I am not sure about its correctness:
- it isn't quite obvious why the `xBounds` and `yBounds` equation above ensures that the `xPos,yPos` computed later will be `< N`. If `xBounds` is a power of 2 and `((centerX - bounds.MinX) / (bounds.MaxX - bounds.MinX))` is 1, then `xPos=N` I think.

- If the bounds are coming from the SRID, I don't see why they can't exceed the size of `uint32` which is undesirable since the `biggestBoxLength` could exceed `uint32`.

I think something like the following would work since it uses the bounds only for normalizing to [0, 1). Once that normalization is done, we can use a constant sized square for the hilbert curve with a side of 2^32.
```
	const boxLength = 1 << 32
	// Add 1 to each bound so that we normalize the coordinates to [0, 1) before
	// multiplying by boxLength to give coordinates that are integers in the interval [0, boxLength-1].
	xBounds := (bounds.MaxX - bounds.MinX) + 1
	yBounds := (bounds.MaxY - bounds.MinY) + 1
	xPos := uint64(((centerX - bounds.MinX) / xBounds) * boxLength)
	yPos := uint64(((centerY - bounds.MinY) / yBounds) * boxLength)
        return hilbertInverse(boxLength, xPos, yPos)
```
</blockquote></details>

Done.
___
*[pkg/geo/geo_test.go, line 565 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDQzwYm0So6gBr2qxdj:-MDRB8z61PKQyrswEFGf:b-iqceia) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d2dd6e52b6b67e75f12201d79f50d702d7a9a4c/pkg/geo/geo_test.go#L565)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why hash?
</blockquote></details>

forgot to rename it.
___
*[pkg/util/encoding/encoding.go, line 982 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDR-uhu19t-b-EYOlfc:-MDR7_hp24cjfA3iNKDn:b-dejz5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d2dd6e52b6b67e75f12201d79f50d702d7a9a4c/pkg/util/encoding/encoding.go#L982)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is this function used for encoding a geometry column even when it is not part of a key (when the ordering does not matter)?
</blockquote></details>

No, that is encoded elsewhere. This is specifically for key encodings which require ordering (indexes & PKs)


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 5 of 5 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-:-MDolTUa5G6PPU8zXtxw:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mjibson, @otan, and @sumeerbhola)

___
*[pkg/geo/geo.go, line 311 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDok7BZ0sYTmAg3sEBI:-MDok7BZ0sYTmAg3sEBJ:b8jqm0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/03a7a375ade666d2e5a07bc83cbb0497d0359298/pkg/geo/geo.go#L311)):*
> ```Go
> 	yBounds := (bounds.MaxY - bounds.MinY) + 1
> 	xPos := uint64(((centerX - bounds.MinX) / xBounds) * boxLength)
> 	yPos := uint64(((centerY - bounds.MinY) / yBounds) * boxLength)
> ```

// hilbertInverse returns values in the interval [0, boxLength^2-1], so [0, 2^64-1]. 
___
*[pkg/geo/geo_test.go, line 641 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDokugT99ShclyYmOEd:-MDokugT99ShclyYmOEe:b-mu1y2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/03a7a375ade666d2e5a07bc83cbb0497d0359298/pkg/geo/geo_test.go#L641)):*
> ```Go
> 				"LINESTRING(0 0, -90 -80)",
> 				"POINT(-80 80)",
> 				"POLYGON((0 0, 1 0, 1 1, 0 1, 0 0))",
> ```

Do you have shapes here that exceed the SRID bounds, and another that is at the bounds, so results in an index of 2^64 -1? Could you add a couple of such cases with assertions on the actual index values (not just relative comparisons)?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51898)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mjibson and @sumeerbhola)

___
*[pkg/geo/geo.go, line 311 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDok7BZ0sYTmAg3sEBI:-MDpL-z26yERe8vuWnom:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03a7a375ade666d2e5a07bc83cbb0497d0359298/pkg/geo/geo.go#L311)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

// hilbertInverse returns values in the interval [0, boxLength^2-1], so [0, 2^64-1]. 
</blockquote></details>

Done.
___
*[pkg/geo/geo_test.go, line 641 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/51898#-MDokugT99ShclyYmOEd:-MDpLe4d2rssbC9x8dR7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03a7a375ade666d2e5a07bc83cbb0497d0359298/pkg/geo/geo_test.go#L641)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Do you have shapes here that exceed the SRID bounds, and another that is at the bounds, so results in an index of 2^64 -1? Could you add a couple of such cases with assertions on the actual index values (not just relative comparisons)?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
geo: order geospatial objects by Hilbert Curve index

This follows the [PostGIS way](https://info.crunchydata.com/blog/waiting-for-postgis-3-hilbert-geometry-sorting),
but does not follow the same encoding mechanism, so we cannot compare
results directly.

For Geography we re-use the S2 infrastructure and for Geometry we use
the google/hilbert library.

Release note (sql change): When ordering by geospatial columns, it will
now order by the Hilbert Space Curve index so that points which are
geographically similar are clustered together.

otan commented:
bors r=sumeerbhola

otan commented:
bors r=sumeerbhola
